### PR TITLE
feat(oh2-287): admin user group crud

### PR DESCRIPTION
**Pull request by gasp**
See oh2-287.

- [x] better handling of user&group tabs (allowing direct access via location state)
- [x] create a new usergroup
- [x] edit usergroups
- [x] manage usergroups permissions
- [x] delete a usergroup

# create a new usergroup
group-create

# edit a usergroup
group-update

# manage usergroup permissions
groups-permissions

# delete a usergroup
group-delete

### CODE REVIEW

**Comment by gasp** [comment]
@mwithi This is using the soon-to-be-deprecated permission api.
Should I disable usergroups' permissions management from this PR?

**Comment by gasp** [comment]
depends on 
https://github.com/informatici/openhospital-api/pull/464 and https://github.com/informatici/openhospital-core/pull/1391

**Comment by mwithi** [comment]
> @mwithi This is using the soon-to-be-deprecated permission api. Should I disable usergroups' permissions management from this PR?

I think you can wait for the API and then adjust this PR, is it ok for you? Thanks!

**Comment by mwithi** [comment]
> depends on [informatici/openhospital-api#464](https://github.com/informatici/openhospital-api/pull/464) and [informatici/openhospital-core#1391](https://github.com/informatici/openhospital-core/pull/1391)

both merged!

File: src/generated/runtime.ts [hunk 1]
```diff
@@ -1,266 +1,211 @@
...
 
-import { Observable, of, Subscriber } from "rxjs";
-import { ajax, AjaxRequest, AjaxResponse } from "rxjs/ajax";
-import { map, concatMap } from "rxjs/operators";
+import { Observable, of, Subscriber } from 'rxjs';
+import { ajax, AjaxRequest, AjaxResponse } from 'rxjs/ajax';
+import { map, concatMap } from 'rxjs/operators';
 
-export const BASE_PATH = "https://oh2.open-hospital.org".replace(/\/+$/, "");
+export const BASE_PATH = 'http://localhost:8080'.replace(/\/+$/, '');
```
Comment (mwithi) [inline src/generated/runtime.ts]: This must remain unchanged @SilverD3 ?

File: src/components/accessories/admin/users/editGroup/EditGroup.tsx [hunk 2]
```diff
@@ -68,46 +82,88 @@ export const EditGroup = () => {
...
-      for (let index = 0; index < permissionsStack.length; index++) {
-        dispatch(
-          updatePermission({
-            id: permissionsStack[index].id,
-            permissionDTO: permissionsStack[index],
-          })
+
+      // this is applying to all permissions, but we may want to only apply to the ones that have changed
+      // using the updatedPermissionsStack array
+      // this is open to debate
```
Comment (mwithi) [inline src/components/accessories/admin/users/editGroup/EditGroup.tsx]: To discuss further

**Comment by mwithi** [comment]
Integration tests seem good, but I would add a "select all" for columns and rows in the permission page.

image

Sorry if I don't remember: the user creation with its password encrypted is part of another issue/PR?

image

**Comment by gasp** [comment]
> Integration tests seem good, but I would add a "select all" for columns and rows in the permission page.
> 
> image
> 

Can we add this as a future iteration? this pull request is already humongous!

**Comment by gasp** [comment]
> Sorry if I don't remember: the user creation with its password encrypted is part of another issue/PR?
> 
> image

Password is related to user, this is usergroup. AFAIK, password encryption is handled backend and does not (should not) affect frontend.

File: src/generated/runtime.ts (hunk 1 above)
Comment (SilverD3) [reply src/generated/runtime.ts]: Yes. See this [comment](https://github.com/informatici/openhospital-ui/pull/662#discussion_r1766295038),

**Comment by SilverD3** [comment]
@gasp I shared a [video on slack](https://openhospitalworkspace.slack.com/archives/CUEL5T15H/p1726816780819749), showing the result on my machine.

1. You can see that the sidebar width grows based on the content of page. You could fix it by controlling the width of the "content display area", I mean the area where we show the content of the selected menu. I remember I've reported that issue in the pass and the fix I applied just worked for the component I was working on.
2. The changes diff you show doesn't work properly. While editing a user group, if check and uncheck the same permission, you'll log it in the changes list (as a permission to revoke)
Screenshot 2024-09-20 at 08-23-29 Open Hospital
3. The reset button doesn't work
4. Maybe you should wait (show a loader) until all the data has been loaded before displaying the edition form. You can see in the video that when I navigate to the user group edit page, some labels show and it's after about 2 or 3 seconds that form fields show. 
5. In the user group edit page, I suggest to use a scrollable component (div) or a collapsible one to display the permission list, so we don't have to scroll too much to reach form actions
6. Until the endpoint for updating a group by passing the list of permissions is improved, you could use what I showed you at the meeting rather than making as many API calls as there are permissions.
```bash
PUT /users/groups/

// Request body
{
  "code": "labo",
  "desc": "Staff members working in the laboratory",
  "permissions": [
    {
      "name": "string",
      "id": 0,
      "description": "string"
    }
  ]
}
```

File: src/components/accessories/admin/users/editGroup/EditGroup.tsx (hunk 2 above)
Comment (mwithi) [reply src/components/accessories/admin/users/editGroup/EditGroup.tsx]: we are going to open an issue for this and, as discussed, the old endpoint should remain active

**Comment by mwithi** [comment]
Please @gasp align with develop and resolve conversations, so we can merge and unlock OH2-399

**Review by SilverD3 — APPROVED** [review]

**Review by mwithi — APPROVED** [review]
Reviewed and approved by @SilverD3

**Pull request merged**